src/shared/wizards/multiStepWizard.ts

```diff
@@ -12,17 +12,74 @@ import * as vscode from 'vscode'
 import * as picker from '../../shared/ui/picker'
 
 export interface WizardStep {
-    (): Thenable<WizardStep | undefined>
+    (): Thenable<Transition>
+}
+
+export interface Transition {
+    nextState: NextWizardState
+    nextStep?: WizardStep
+}
+
+export enum NextWizardState {
+    CONTINUE,
+    REPROMPT,
+    GO_BACK,
+    TERMINATE,
+}
+
+export const WIZARD_REPROMPT: Transition = {
+    nextState: NextWizardState.REPROMPT,
+}
+
+export const WIZARD_TERMINATE: Transition = {
+    nextState: NextWizardState.TERMINATE,
+}
+
+export const WIZARD_GOBACK: Transition = {
+    nextState: NextWizardState.GO_BACK,
+}
+
+export function wizardContinue(step: WizardStep): Transition {
+    return {
+        nextState: NextWizardState.CONTINUE,
+        nextStep: step,
+    }
 }
 
 export abstract class MultiStepWizard<TResult> {
     protected constructor() {}
 
     public async run(): Promise<TResult | undefined> {
-        let step: WizardStep | undefined = this.startStep
-
-        while (step) {
-            step = await step()
+        let steps: WizardStep[] = [this.startStep]
+
+        // in a wizard, it only make sense to go forwards to an arbitrary step, or backwards in history
+        while (steps.length > 0) {
+            const step = steps[steps.length - 1]
+            // non-terminal if we still have steps
+            if (!step) {
+                break
```

Comment:
This infinite loops if you access an `undefined` object at the top of the stack. (as an example, Line 78 could inadvertently add an undefined entry to `steps`)

Reply:
Seems like we are terminating in the face of an unexpected undefined value. Consider logging here.

Another option would be to pop the undefined and keep looping. Is it worth throwing an error if this is an unexpected event?

Reply:
added a log, but i feel like if we keep looping that would be more confusing than exiting.